rasterio/_base.pyx

```diff
@@ -164,7 +164,10 @@ cdef class DatasetBase(object):
 
     cdef GDALDatasetH handle(self) except NULL:
         """Return the object's GDAL dataset handle"""
-        return self._hds
+        if self._hds == NULL:
```

Comment:
Doing something like:

```
with rasterio.open('tests/data/RGB.byte.tif') as src:
    pass
src.files
```

would raise a not useful `SystemError`.  Most (all?) of the other properties and attributes are cached on first read, but there are situations where `files` may be modified while the dataset is open.  Technically we could cache on close, but I don't think we want users relying on getting information about a dataset when its not open.